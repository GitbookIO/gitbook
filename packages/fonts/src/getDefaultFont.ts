import type { CustomizationDefaultFont } from '@gitbook/api';
import { fonts } from './fonts';
import type { FontWeight } from './types';

/**
 * Get the URL to load a font for a text.
 */
export function getDefaultFont(input: {
    /**
     * GitBook font to use.
     */
    font: CustomizationDefaultFont;

    /**
     * Text to display with the font.
     */
    text: string;

    /**
     * Font weight to use.
     */
    weight: FontWeight;
}): { font: string; url: string } | null {
    if (!input.text.trim()) {
        return null;
    }

    const fontDefinition = fonts[input.font];
    if (!fontDefinition) {
        return null;
    }

    const variant = fontDefinition.variants[`${input.weight}`];
    if (!variant) {
        return null;
    }

    const script = getBestUnicodeRange(input.text, fontDefinition.unicodeRange);
    if (!script) {
        return null;
    }

    return variant[script]
        ? {
              font: input.font,
              url: variant[script],
          }
        : null;
}

/**
 * Determine which named @font-face unicode-range covers
 * the greatest share of the characters in `text`.
 *
 * @param text   The text you want to inspect.
 * @param ranges An object whose keys are range names and whose
 *               values are CSS-style comma-separated unicode-range
 *               declarations (e.g. "U+0370-03FF,U+1F00-1FFF").
 * @returns      The key of the best-matching range, or `null`
 *               when nothing matches at all.
 */
function getBestUnicodeRange(text: string, ranges: Record<string, string>): string | null {
    // ---------- helper: parse "U+XXXX" or "U+XXXX-YYYY" ----------
    const parseOne = (token: string): [number, number] | null => {
        token = token.trim().toUpperCase();
        if (!token.startsWith('U+')) return null;

        const body = token.slice(2); // drop "U+"
        const [startHex, endHex] = body.split('-');
        const start = Number.parseInt(startHex!, 16);
        const end = endHex ? Number.parseInt(endHex, 16) : start;

        if (Number.isNaN(start) || Number.isNaN(end) || end < start) return null;
        return [start, end];
    };

    // ---------- helper: build lookup table ----------
    const parsed: Record<string, [number, number][]> = {};
    for (const [label, list] of Object.entries(ranges)) {
        parsed[label] = list
            .split(',')
            .map(parseOne)
            .filter((x): x is [number, number] => x !== null);
    }

    // ---------- tally code-point hits ----------
    const hits: Record<string, number> = Object.fromEntries(Object.keys(parsed).map((k) => [k, 0]));

    for (let i = 0; i < text.length; ) {
        const cp = text.codePointAt(i)!;
        i += cp > 0xffff ? 2 : 1; // advance by 1 UTF-16 code-unit (or 2 for surrogates)

        for (const [label, rangesArr] of Object.entries(parsed)) {
            if (rangesArr.some(([lo, hi]) => cp >= lo && cp <= hi)) {
                hits[label]!++;
            }
        }
    }

    // ---------- choose the "best" ----------
    let winner: string | null = null;
    let maxCount = 0;

    for (const [label, count] of Object.entries(hits)) {
        if (count > maxCount) {
            maxCount = count;
            winner = label;
        }
    }

    return maxCount > 0 ? winner : null;
}
