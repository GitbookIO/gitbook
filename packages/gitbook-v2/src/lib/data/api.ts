import { trace } from '@/lib/tracing';
import {
    type ComputedContentSource,
    GitBookAPI,
    type GitBookAPIServiceBinding,
    type RenderIntegrationUI,
} from '@gitbook/api';
import { getCacheTag, getComputedContentSourceCacheTags } from '@gitbook/cache-tags';
import { GITBOOK_API_TOKEN, GITBOOK_API_URL, GITBOOK_USER_AGENT } from '@v2/lib/env';
import { unstable_cacheLife as cacheLife, unstable_cacheTag as cacheTag } from 'next/cache';
import { DataFetcherError, wrapDataFetcherError } from './errors';
import { memoize } from './memoize';
import type { GitBookDataFetcher } from './types';

interface DataFetcherInput {
    /**
     * API token.
     */
    apiToken: string | null;
}

/**
 * Create a data fetcher using an API token.
 * The data are being cached by Next.js built-in cache.
 */
export function createDataFetcher(
    input: DataFetcherInput = { apiToken: null }
): GitBookDataFetcher {
    return {
        async api() {
            return apiClient(input);
        },

        withToken({ apiToken }) {
            return createDataFetcher({
                apiToken,
            });
        },

        //
        // API that are tied to the token
        //
        getPublishedContentSite(params) {
            return trace('getPublishedContentSite', () =>
                getPublishedContentSite(input, {
                    organizationId: params.organizationId,
                    siteId: params.siteId,
                    siteShareKey: params.siteShareKey,
                })
            );
        },
        getSiteRedirectBySource(params) {
            return trace('getSiteRedirectBySource', () =>
                getSiteRedirectBySource(input, {
                    organizationId: params.organizationId,
                    siteId: params.siteId,
                    siteShareKey: params.siteShareKey,
                    source: params.source,
                })
            );
        },
        getRevision(params) {
            return trace('getRevision', () =>
                getRevision(input, {
                    spaceId: params.spaceId,
                    revisionId: params.revisionId,
                    metadata: params.metadata,
                })
            );
        },
        getRevisionPages(params) {
            return trace('getRevisionPages', () =>
                getRevisionPages(input, {
                    spaceId: params.spaceId,
                    revisionId: params.revisionId,
                    metadata: params.metadata,
                })
            );
        },
        getRevisionFile(params) {
            return trace('getRevisionFile', () =>
                getRevisionFile(input, {
                    spaceId: params.spaceId,
                    revisionId: params.revisionId,
                    fileId: params.fileId,
                })
            );
        },
        getRevisionPageByPath(params) {
            return trace('getRevisionPageByPath', () =>
                getRevisionPageByPath(input, {
                    spaceId: params.spaceId,
                    revisionId: params.revisionId,
                    path: params.path,
                })
            );
        },
        getRevisionPageMarkdown(params) {
            return trace('getRevisionPageMarkdown', () =>
                getRevisionPageMarkdown(input, {
                    spaceId: params.spaceId,
                    revisionId: params.revisionId,
                    pageId: params.pageId,
                })
            );
        },
        getReusableContent(params) {
            return trace('getReusableContent', () =>
                getReusableContent(input, {
                    spaceId: params.spaceId,
                    revisionId: params.revisionId,
                    reusableContentId: params.reusableContentId,
                })
            );
        },
        getLatestOpenAPISpecVersionContent(params) {
            return trace('getLatestOpenAPISpecVersionContent', () =>
                getLatestOpenAPISpecVersionContent(input, {
                    organizationId: params.organizationId,
                    slug: params.slug,
                })
            );
        },
        getSpace(params) {
            return trace('getSpace', () =>
                getSpace(input, {
                    spaceId: params.spaceId,
                    shareKey: params.shareKey,
                })
            );
        },
        getChangeRequest(params) {
            return trace('getChangeRequest', () =>
                getChangeRequest(input, {
                    spaceId: params.spaceId,
                    changeRequestId: params.changeRequestId,
                })
            );
        },
        getDocument(params) {
            return trace('getDocument', () =>
                getDocument(input, {
                    spaceId: params.spaceId,
                    documentId: params.documentId,
                })
            );
        },
        getComputedDocument(params) {
            return trace('getComputedDocument', () =>
                getComputedDocument(input, {
                    organizationId: params.organizationId,
                    spaceId: params.spaceId,
                    source: params.source,
                    seed: params.seed,
                })
            );
        },
        getEmbedByUrl(params) {
            return trace('getEmbedByUrl', () =>
                getEmbedByUrl(input, {
                    url: params.url,
                    spaceId: params.spaceId,
                })
            );
        },
        searchSiteContent(params) {
            return trace('searchSiteContent', () => searchSiteContent(input, params));
        },

        renderIntegrationUi(params) {
            return trace('renderIntegrationUi', () =>
                renderIntegrationUi(input, {
                    integrationName: params.integrationName,
                    request: params.request,
                })
            );
        },

        getUserById(userId) {
            return trace('getUserById', () => getUserById(input, { userId }));
        },
    };
}

const getUserById = memoize(async function getUserById(
    input: DataFetcherInput,
    params: { userId: string }
) {
    'use cache';

    return trace('getUserById.uncached', () => {
        cacheLife('days');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.users.getUserById(params.userId);
            return res.data;
        });
    });
});

const getSpace = memoize(async function getSpace(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        shareKey: string | undefined;
    }
) {
    'use cache';

    return trace('getSpace.uncached', () => {
        cacheLife('days');
        cacheTag(
            getCacheTag({
                tag: 'space',
                space: params.spaceId,
            })
        );

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getSpaceById(params.spaceId, {
                shareKey: params.shareKey,
            });
            return res.data;
        });
    });
});

const getChangeRequest = memoize(async function getChangeRequest(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        changeRequestId: string;
    }
) {
    'use cache';

    return trace('getChangeRequest.uncached', () => {
        cacheLife('minutes');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getChangeRequestById(
                params.spaceId,
                params.changeRequestId
            );
            cacheTag(
                getCacheTag({
                    tag: 'change-request',
                    space: params.spaceId,
                    changeRequest: res.data.id,
                })
            );
            return res.data;
        });
    });
});

const getRevision = memoize(async function getRevision(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        revisionId: string;
        metadata: boolean;
    }
) {
    'use cache';

    return trace('getRevision.uncached', () => {
        cacheLife('max');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getRevisionById(params.spaceId, params.revisionId, {
                metadata: params.metadata,
            });
            return res.data;
        });
    });
});

const getRevisionPages = memoize(async function getRevisionPages(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        revisionId: string;
        metadata: boolean;
    }
) {
    'use cache';

    return trace('getRevisionPages.uncached', () => {
        cacheLife('max');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.listPagesInRevisionById(
                params.spaceId,
                params.revisionId,
                {
                    metadata: params.metadata,
                }
            );
            return res.data.pages;
        });
    });
});

const getRevisionFile = memoize(async function getRevisionFile(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        revisionId: string;
        fileId: string;
    }
) {
    'use cache';

    return trace('getRevisionFile.uncached', () => {
        cacheLife('max');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getFileInRevisionById(
                params.spaceId,
                params.revisionId,
                params.fileId,
                {}
            );
            return res.data;
        });
    });
});

const getRevisionPageMarkdown = memoize(async function getRevisionPageMarkdown(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        revisionId: string;
        pageId: string;
    }
) {
    'use cache';

    return trace('getRevisionPageMarkdown.uncached', () => {
        cacheLife('max');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getPageInRevisionById(
                params.spaceId,
                params.revisionId,
                params.pageId,
                {
                    format: 'markdown',
                }
            );

            if (!('markdown' in res.data)) {
                throw new DataFetcherError('Page is not a document', 404);
            }

            return res.data.markdown;
        });
    });
});

const getRevisionPageByPath = memoize(async function getRevisionPageByPath(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        revisionId: string;
        path: string;
    }
) {
    'use cache';

    return trace('getRevisionPageByPath.uncached', () => {
        cacheLife('max');

        const encodedPath = encodeURIComponent(params.path);
        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getPageInRevisionByPath(
                params.spaceId,
                params.revisionId,
                encodedPath,
                {}
            );

            return res.data;
        });
    });
});

const getDocument = memoize(async function getDocument(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        documentId: string;
    }
) {
    'use cache';

    return trace('getDocument.uncached', () => {
        cacheLife('max');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getDocumentById(params.spaceId, params.documentId, {});
            return res.data;
        });
    });
});

const getComputedDocument = memoize(async function getComputedDocument(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        organizationId: string;
        source: ComputedContentSource;
        seed: string;
    }
) {
    'use cache';

    return trace('getComputedDocument.uncached', () => {
        cacheLife('days');

        cacheTag(
            ...getComputedContentSourceCacheTags(
                {
                    spaceId: params.spaceId,
                    organizationId: params.organizationId,
                },
                params.source
            )
        );

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getComputedDocument(params.spaceId, {
                source: params.source,
                seed: params.seed,
            });
            return res.data;
        });
    });
});

const getReusableContent = memoize(async function getReusableContent(
    input: DataFetcherInput,
    params: {
        spaceId: string;
        revisionId: string;
        reusableContentId: string;
    }
) {
    'use cache';

    return trace('getReusableContent.uncached', () => {
        cacheLife('max');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getReusableContentInRevisionById(
                params.spaceId,
                params.revisionId,
                params.reusableContentId
            );
            return res.data;
        });
    });
});

const getLatestOpenAPISpecVersionContent = memoize(
    async function getLatestOpenAPISpecVersionContent(
        input: DataFetcherInput,
        params: {
            organizationId: string;
            slug: string;
        }
    ) {
        'use cache';

        return trace('getLatestOpenAPISpecVersionContent.uncached', () => {
            cacheTag(
                getCacheTag({
                    tag: 'openapi',
                    organization: params.organizationId,
                    openAPISpec: params.slug,
                })
            );
            cacheLife('days');

            return wrapDataFetcherError(async () => {
                const api = await apiClient(input);
                const res = await api.orgs.getLatestOpenApiSpecVersionContent(
                    params.organizationId,
                    params.slug
                );
                return res.data;
            });
        });
    }
);

const getPublishedContentSite = memoize(async function getPublishedContentSite(
    input: DataFetcherInput,
    params: {
        organizationId: string;
        siteId: string;
        siteShareKey: string | undefined;
    }
) {
    'use cache';

    return trace('getPublishedContentSite.uncached', () => {
        cacheLife('days');
        cacheTag(
            getCacheTag({
                tag: 'site',
                site: params.siteId,
            })
        );

        return trace('getPublishedContentSite', () => {
            return wrapDataFetcherError(async () => {
                const api = await apiClient(input);
                const res = await api.orgs.getPublishedContentSite(
                    params.organizationId,
                    params.siteId,
                    {
                        shareKey: params.siteShareKey,
                    }
                );
                return res.data;
            });
        });
    });
});

const getSiteRedirectBySource = memoize(async function getSiteRedirectBySource(
    input: DataFetcherInput,
    params: {
        organizationId: string;
        siteId: string;
        siteShareKey: string | undefined;
        source: string;
    }
) {
    'use cache';

    return trace('getSiteRedirectBySource.uncached', () => {
        cacheTag(
            getCacheTag({
                tag: 'site',
                site: params.siteId,
            })
        );
        cacheLife('days');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.orgs.getSiteRedirectBySource(
                params.organizationId,
                params.siteId,
                {
                    shareKey: params.siteShareKey,
                    source: params.source,
                }
            );

            return res.data;
        });
    });
});

const getEmbedByUrl = memoize(async function getEmbedByUrl(
    input: DataFetcherInput,
    params: {
        url: string;
        spaceId: string;
    }
) {
    'use cache';

    return trace('getEmbedByUrl.uncached', () => {
        cacheLife('weeks');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.spaces.getEmbedByUrlInSpace(params.spaceId, { url: params.url });
            return res.data;
        });
    });
});

const searchSiteContent = memoize(async function searchSiteContent(
    input: DataFetcherInput,
    params: Parameters<GitBookDataFetcher['searchSiteContent']>[0]
) {
    'use cache';

    return trace('searchSiteContent.uncached', () => {
        const { organizationId, siteId, query, scope } = params;

        cacheLife('days');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.orgs.searchSiteContent(organizationId, siteId, {
                query,
                ...scope,
            });
            return res.data.items;
        });
    });
});

const renderIntegrationUi = memoize(async function renderIntegrationUi(
    input: DataFetcherInput,
    params: {
        integrationName: string;
        request: RenderIntegrationUI;
    }
) {
    'use cache';

    return trace('renderIntegrationUi.uncached', () => {
        cacheTag(getCacheTag({ tag: 'integration', integration: params.integrationName }));
        cacheLife('days');

        return wrapDataFetcherError(async () => {
            const api = await apiClient(input);
            const res = await api.integrations.renderIntegrationUiWithPost(
                params.integrationName,
                params.request
            );
            return res.data;
        });
    });
});

let loggedServiceBinding = false;

/**
 * Create a new API client.
 */
export async function apiClient(input: DataFetcherInput = { apiToken: null }) {
    const { apiToken } = input;
    let serviceBinding: GitBookAPIServiceBinding | undefined;

    try {
        // HACK: This is a workaround to avoid webpack trying to bundle this cloudflare only module
        // @ts-ignore
        const { env } = await import(
            /* webpackIgnore: true */ `${'__cloudflare:workers'.replaceAll('_', '')}`
        );
        serviceBinding = env.GITBOOK_API;
        if (!loggedServiceBinding) {
            loggedServiceBinding = true;
            if (serviceBinding) {
                // biome-ignore lint/suspicious/noConsole: we want to log here
                console.log(`using service binding for the API (${GITBOOK_API_URL})`);
            } else {
                // biome-ignore lint/suspicious/noConsole: we want to log here
                console.warn(`no service binding for the API (${GITBOOK_API_URL})`);
            }
        }
    } catch (error) {
        if (process.env.NODE_ENV === 'production' && !process.env.VERCEL) {
            throw error;
        }
    }

    const api = new GitBookAPI({
        authToken: apiToken || GITBOOK_API_TOKEN || undefined,
        endpoint: GITBOOK_API_URL,
        userAgent: GITBOOK_USER_AGENT,
        serviceBinding,
    });

    return api;
}
