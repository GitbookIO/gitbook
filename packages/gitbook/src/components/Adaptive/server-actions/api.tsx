'use server';
import {
    type AIMessage,
    AIMessageRole,
    type AIMessageStep,
    type AIStreamResponse,
} from '@gitbook/api';
import type { GitBookBaseContext } from '@v2/lib/context';
import type { GitBookDataFetcher } from '@v2/lib/data';
import { EventIterator } from 'event-iterator';
import type { MaybePromise } from 'p-map';
import * as partialJson from 'partial-json';
import type { DeepPartial } from 'ts-essentials';
import type { z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';
import { MessageView } from '../MessageView';

/**
 * Get the latest value from a stream and the response id.
 */
export async function generate<T>(
    promise: MaybePromise<{
        stream: EventIterator<T>;
        response: Promise<{ responseId: string }>;
    }>
) {
    const input = await promise;
    let value: T | undefined;

    for await (const event of input.stream) {
        value = event;
    }

    const { responseId } = await input.response;
    return {
        responseId,
        value,
    };
}

/**
 * Stream the generation of an object using the AI.
 */
export async function streamGenerateObject<T>(
    context: GitBookBaseContext,
    {
        schema,
        ...input
    }: Omit<Parameters<GitBookDataFetcher['streamAIResponse']>[0], 'output'> & {
        schema: z.ZodSchema<T>;
    }
) {
    const rawStream = context.dataFetcher.streamAIResponse({
        ...input,
        output: {
            type: 'object',
            schema: zodToJsonSchema(schema),
        },
    });

    let json = '';
    return parseResponse<DeepPartial<T>>(rawStream, (event) => {
        if (event.type === 'response_object') {
            json += event.jsonChunk;

            const parsed = partialJson.parse(json, partialJson.ALL);
            return parsed;
        }
    });
}

/**
 * Stream the generation of a document.
 */
export async function streamGenerateDocument(
    context: GitBookBaseContext,
    input: Omit<Parameters<GitBookDataFetcher['streamAIResponse']>[0], 'output'>
) {
    const rawStream = context.dataFetcher.streamAIResponse({
        ...input,
        output: {
            type: 'document',
        },
    });

    const message: AIMessage = {
        id: '',
        role: AIMessageRole.Assistant,
        steps: [],
    };

    const updateProcessingMessageStep = (
        stepIndex: number,
        callback: (step: AIMessageStep) => void
    ) => {
        if (stepIndex > message.steps.length) {
            throw new Error(
                `Step index out of bounds ${stepIndex} (${message.steps.length} steps)`
            );
        }

        if (message.steps[stepIndex]) {
            message.steps = [...message.steps];
            message.steps[stepIndex] = { ...message.steps[stepIndex] };
            callback(message.steps[stepIndex]);
        } else {
            message.steps = [
                ...message.steps,
                {
                    content: {
                        object: 'document',
                        data: {},
                        nodes: [],
                    },
                },
            ];
            callback(message.steps[stepIndex]);
        }
    };

    return parseResponse<React.ReactNode>(rawStream, (event) => {
        switch (event.type) {
            /**
             * The agent is processing a tool call in a new message.
             */
            case 'response_tool_call': {
                updateProcessingMessageStep(event.stepIndex, (step) => {
                    step.toolCalls ??= [];
                    step.toolCalls.push(event.toolCall);
                });
                break;
            }

            /**
             * The agent is writing the content of a new message.
             */
            case 'response_reasoning':
            case 'response_document': {
                updateProcessingMessageStep(event.stepIndex, (step) => {
                    const container = event.type === 'response_reasoning' ? 'reasoning' : 'content';

                    step[container] ??= {
                        object: 'document',
                        data: {},
                        nodes: [],
                    };
                    step[container] = {
                        ...step[container],
                        nodes: [...step[container].nodes],
                    };
                    if (event.operation === 'insert') {
                        step[container].nodes.push(...event.blocks);
                    } else {
                        step[container].nodes.splice(
                            -event.blocks.length,
                            event.blocks.length,
                            ...event.blocks
                        );
                    }
                });
                break;
            }
        }

        return <MessageView message={message} />;
    });
}

/**
 * Parse a stream from the API to extract the responseId.
 */
function parseResponse<T>(
    responseStream: EventIterator<AIStreamResponse>,
    parse: (response: AIStreamResponse) => T | undefined
): {
    stream: EventIterator<T>;
    response: Promise<{ responseId: string }>;
} {
    let resolveResponse: (value: { responseId: string }) => void;
    const response = new Promise<{ responseId: string }>((resolve) => {
        resolveResponse = resolve;
    });

    const stream = new EventIterator<T>((queue) => {
        (async () => {
            let foundResponse = false;

            for await (const event of responseStream) {
                if (event.type === 'response_finish') {
                    foundResponse = true;
                    resolveResponse({ responseId: event.responseId });
                } else {
                    const parsed = parse(event);
                    if (parsed !== undefined) {
                        queue.push(parsed);
                    }
                }
            }

            if (!foundResponse) {
                throw new Error('No response found');
            }
        })().then(
            () => {
                queue.stop();
            },
            (error) => {
                queue.fail(error);
            }
        );
    });

    return { stream, response };
}
