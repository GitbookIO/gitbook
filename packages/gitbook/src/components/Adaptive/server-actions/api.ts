'use server';
import { type AIMessageInput, AIModel, type AIStreamResponse } from '@gitbook/api';
import type { GitBookBaseContext } from '@v2/lib/context';
import { EventIterator } from 'event-iterator';
import type { MaybePromise } from 'p-map';
import * as partialJson from 'partial-json';
import type { DeepPartial } from 'ts-essentials';
import type { z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

/**
 * Get the latest value from a stream and the response id.
 */
export async function generate<T>(
    promise: MaybePromise<{
        stream: EventIterator<T>;
        response: Promise<{ responseId: string }>;
    }>
) {
    const input = await promise;
    let value: T | undefined;

    for await (const event of input.stream) {
        value = event;
    }

    const { responseId } = await input.response;
    return {
        responseId,
        value,
    };
}

/**
 * Stream the generation of an object using the AI.
 */
export async function streamGenerateObject<T>(
    context: GitBookBaseContext,
    {
        organizationId,
        siteId,
    }: {
        organizationId: string;
        siteId: string;
    },
    {
        schema,
        messages,
        model = AIModel.Fast,
    }: {
        schema: z.ZodSchema<T>;
        messages: AIMessageInput[];
        model?: AIModel;
        previousResponseId?: string;
    }
) {
    const rawStream = context.dataFetcher.streamAIResponse({
        organizationId,
        siteId,
        input: messages,
        output: {
            type: 'object',
            schema: zodToJsonSchema(schema),
        },
        model,
    });

    let json = '';
    return parseResponse<DeepPartial<T>>(rawStream, (event) => {
        if (event.type === 'response_object') {
            json += event.jsonChunk;

            const parsed = partialJson.parse(json, partialJson.ALL);
            return parsed;
        }
    });
}

/**
 * Parse a stream from the API to extract the responseId.
 */
function parseResponse<T>(
    responseStream: EventIterator<AIStreamResponse>,
    parse: (response: AIStreamResponse) => T | undefined
): {
    stream: EventIterator<T>;
    response: Promise<{ responseId: string }>;
} {
    let resolveResponse: (value: { responseId: string }) => void;
    const response = new Promise<{ responseId: string }>((resolve) => {
        resolveResponse = resolve;
    });

    const stream = new EventIterator<T>((queue) => {
        (async () => {
            let foundResponse = false;

            for await (const event of responseStream) {
                if (event.type === 'response_finish') {
                    foundResponse = true;
                    resolveResponse({ responseId: event.responseId });
                } else {
                    const parsed = parse(event);
                    if (parsed !== undefined) {
                        queue.push(parsed);
                    }
                }
            }

            if (!foundResponse) {
                throw new Error('No response found');
            }
        })().then(
            () => {
                queue.stop();
            },
            (error) => {
                queue.fail(error);
            }
        );
    });

    return { stream, response };
}
