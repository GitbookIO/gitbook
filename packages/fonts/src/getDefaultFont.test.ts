import { describe, expect, it } from 'bun:test';
import { CustomizationDefaultFont } from '@gitbook/api';
import { getDefaultFont } from './getDefaultFont';

describe('getDefaultFont', () => {
    it('should return null for invalid font', () => {
        const result = getDefaultFont({
            font: 'invalid-font' as CustomizationDefaultFont,
            text: 'Hello',
            weight: 400,
        });
        expect(result).toBeNull();
    });

    it('should return null for invalid weight', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Hello',
            weight: 999 as any,
        });
        expect(result).toBeNull();
    });

    it('should return null for text not supported by any script', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'ðŸ˜€', // Emoji not supported by Inter
            weight: 400,
        });
        expect(result).toBeNull();
    });

    it('should return correct object for Latin text', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Hello World',
            weight: 400,
        });
        expect(result).not.toBeNull();
        expect(result?.font).toBe(CustomizationDefaultFont.Inter);
        expect(result).toMatchSnapshot();
    });

    it('should return correct object for Cyrillic text', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'ÐŸÑ€Ð¸Ð²ÐµÑ‚ Ð¼Ð¸Ñ€',
            weight: 400,
        });
        expect(result).not.toBeNull();
        expect(result?.font).toBe(CustomizationDefaultFont.Inter);
        expect(result).toMatchSnapshot();
    });

    it('should return correct object for Greek text', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Î“ÎµÎ¹Î± ÏƒÎ±Ï‚',
            weight: 400,
        });
        expect(result).not.toBeNull();
        expect(result?.font).toBe(CustomizationDefaultFont.Inter);
        expect(result).toMatchSnapshot();
    });

    it('should handle mixed script text', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Hello ÐŸÑ€Ð¸Ð²ÐµÑ‚',
            weight: 400,
        });
        expect(result).not.toBeNull();
        expect(result?.font).toBe(CustomizationDefaultFont.Inter);
        expect(result).toMatchSnapshot();
    });

    it('should handle different font weights', () => {
        const regular = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Hello',
            weight: 400,
        });
        const bold = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Hello',
            weight: 700,
        });
        expect(regular).not.toBeNull();
        expect(bold).not.toBeNull();
        expect(regular).toMatchSnapshot('regular');
        expect(bold).toMatchSnapshot('bold');
    });

    it('should handle empty string', () => {
        const result = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: '',
            weight: 400,
        });
        expect(result).toBeNull();
    });

    it('should handle different fonts', () => {
        const inter = getDefaultFont({
            font: CustomizationDefaultFont.Inter,
            text: 'Hello',
            weight: 400,
        });
        const roboto = getDefaultFont({
            font: CustomizationDefaultFont.Roboto,
            text: 'Hello',
            weight: 400,
        });
        expect(inter).not.toBeNull();
        expect(roboto).not.toBeNull();
        expect(inter).toMatchSnapshot('inter');
        expect(roboto).toMatchSnapshot('roboto');
    });
});
