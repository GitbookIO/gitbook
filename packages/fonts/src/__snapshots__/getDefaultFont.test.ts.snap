// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getDefaultFont should return correct object for Latin text 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfAZ9hjp-Ek-_0ew.woff",
}
`;

exports[`getDefaultFont should return correct object for Cyrillic text 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfAZthjp-Ek-_0ewmM.woff",
}
`;

exports[`getDefaultFont should return correct object for Greek text 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfAZxhjp-Ek-_0ewmM.woff",
}
`;

exports[`getDefaultFont should handle mixed script text 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfAZthjp-Ek-_0ewmM.woff",
}
`;

exports[`getDefaultFont should handle different font weights: regular 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfAZ9hjp-Ek-_0ew.woff",
}
`;

exports[`getDefaultFont should handle different font weights: bold 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuFuYAZ9hjp-Ek-_0ew.woff",
}
`;

exports[`getDefaultFont should handle different fonts: inter 1`] = `
{
  "font": "Inter",
  "url": "https://fonts.gstatic.com/s/inter/v18/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfAZ9hjp-Ek-_0ew.woff",
}
`;

exports[`getDefaultFont should handle different fonts: roboto 1`] = `
{
  "font": "Roboto",
  "url": "https://fonts.gstatic.com/s/roboto/v32/KFOmCnqEu92Fr1Mu4mxMKTU1Kg.woff",
}
`;
